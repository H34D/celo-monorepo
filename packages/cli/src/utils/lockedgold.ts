export const LockedGoldArgs = {
  pendingWithdrawalIndexArg: {
    name: 'pendingWithdrawalINdex',
    description: 'index of pending withdrawal whose unlocking period has passed',
  },
  goldAmountArg: {
    name: 'goldAmount',
    description: 'unit amount of gold token (cGLD)',
  },
}
