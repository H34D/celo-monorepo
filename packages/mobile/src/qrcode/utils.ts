import { isValidAddress } from '@celo/utils/src/signatureUtils'
import { isEmpty } from 'lodash'
import * as RNFS from 'react-native-fs'
import Share from 'react-native-share'
import { all, put } from 'redux-saga/effects'
import { showError } from 'src/alert/actions'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { updateE164PhoneNumberAddresses } from 'src/identity/actions'
import { AddressToE164NumberType } from 'src/identity/reducer'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { setRecipientCache } from 'src/recipients/actions'
import {
  getRecipientFromAddress,
  NumberToRecipient,
  Recipient,
  RecipientKind,
} from 'src/recipients/recipient'
import { QrCode, storeLatestInRecents, SVG } from 'src/send/actions'
import Logger from 'src/utils/Logger'
import uuidv1 from 'uuid'

const TAG = 'QR/utils'

export enum BarcodeTypes {
  QR_CODE = 'QR_CODE',
}

const QRFileName = '/celo-qr.png'

export async function shareSVGImage(svg: SVG) {
  if (!svg) {
    return
  }
  svg.toDataURL(async (data: string) => {
    const path = RNFS.DocumentDirectoryPath + QRFileName
    try {
      await RNFS.writeFile(path, data, 'base64')
      Share.open({
        url: 'file://' + path,
        type: 'image/png',
      }).catch((err: Error) => {
        throw err
      })
    } catch (e) {
      Logger.warn(TAG, e)
    }
  })
}

export function* handleBarcode(
  barcode: QrCode,
  addressToE164Number: AddressToE164NumberType,
  recipientCache: NumberToRecipient
) {
  if (barcode.type !== BarcodeTypes.QR_CODE) {
    return
  }

  let data: { address: string; e164PhoneNumber: string; displayName: string } | undefined
  try {
    data = JSON.parse(barcode.data)
  } catch (e) {
    Logger.warn(TAG, 'QR code read failed with ' + e)
  }

  if (typeof data !== 'object' || isEmpty(data.address)) {
    yield put(showError(ErrorMessages.QR_FAILED_NO_ADDRESS))
    return
  }
  if (!isValidAddress(data.address)) {
    yield put(showError(ErrorMessages.QR_FAILED_INVALID_ADDRESS))
    return
  }
  if (typeof data.e164PhoneNumber !== 'string') {
    // Default for invalid e164PhoneNumber
    data.e164PhoneNumber = ''
  }
  if (typeof data.displayName !== 'string') {
    // Default for invalid displayName
    data.displayName = ''
  }

  const cachedRecipient = getRecipientFromAddress(data.address, addressToE164Number, recipientCache)

  const recipient: Recipient = cachedRecipient
    ? {
        ...data,
        kind: RecipientKind.QrCode,
        displayId: data.e164PhoneNumber,
        phoneNumberLabel: 'QR Code',
        thumbnailPath: cachedRecipient.thumbnailPath,
        contactId: cachedRecipient.contactId,
      }
    : {
        ...data,
        kind: RecipientKind.QrCode,
        displayId: data.e164PhoneNumber,
      }

  // TODO data in the QR should be encrypted so no one can create an arbitrary QR with any phone and address combination
  // add phone number to the recipientCache
  recipientCache[data.e164PhoneNumber] = {
    ...recipient,
    kind: RecipientKind.Contact,
    contactId: recipient.contactId ? recipient.contactId : uuidv1(),
  }
  addressToE164Number[data.address] = data.e164PhoneNumber

  yield all([
    put(setRecipientCache(recipientCache)),
    put(storeLatestInRecents(recipient)),
    put(updateE164PhoneNumberAddresses({}, addressToE164Number)),
  ])

  navigate(Screens.SendAmount, { recipient })
}
