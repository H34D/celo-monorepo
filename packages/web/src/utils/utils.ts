import NetworkSpeed from 'network-speed'

export function randomIntegerInRange(min: number, max: number) {
  return Math.round(Math.random() * (max - min + 1)) + min
}

export function scrollTo(elementID: string, position?: 'start' | 'center') {
  const element = document.getElementById(elementID)
  if (element) {
    element.scrollIntoView({
      behavior: 'smooth',
      block: position,
      inline: position,
    })
  }
}

interface Speeds {
  mbps: string
  kbps: string
  bps: string
}

async function getNetworkDownloadSpeed() {
  const testNetworkSpeed = new NetworkSpeed()
  const baseUrl = 'http://httpbin.org/bytes/10000000'
  const fileSize = 100000
  const speed: Speeds = await testNetworkSpeed.checkDownloadSpeed(baseUrl, fileSize)
  return isFast(Number(speed.mbps))
}

const MIN_MB_FOR_FAST = 3

async function isFast(speed: number | EffectiveTypes) {
  if (speed === EffectiveTypes['4g']) {
    return true
  }
  if (typeof speed === 'number' && speed > MIN_MB_FOR_FAST) {
    return true
  }
  return false
}

function getEffectiveConnection(navigator): EffectiveTypes {
  const connection = navigator.connection || navigator.mozConnection || navigator.webkitConnection
  return (connection && connection.effectiveType) || 'unknown'
}

// from http://wicg.github.io/netinfo/#dom-effectiveconnectiontype-slow-2g
enum EffectiveTypes {
  '2g' = '2g',
  '3g' = '3g',
  '4g' = '4g',
  'slow-2g' = 'slow-2g',
  'unknown' = 'unknown',
}

export async function hasGoodConnection() {
  const chromesBuiltInMethod = getEffectiveConnection(window.navigator)
  if (chromesBuiltInMethod !== 'unknown') {
    return isFast(chromesBuiltInMethod)
  }

  return Promise.race([getNetworkDownloadSpeed(), abort()])
}

const MAX_TIME_MS = 3000

async function abort(): Promise<boolean> {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(false)
    }, MAX_TIME_MS)
  })
}

type MemoryGB = 0.25 | 0.5 | 1 | 2 | 4 | 8

export function getDeviceMemory(): MemoryGB {
  // only available on chrome / android browser assume 4 if we dont know
  // @ts-ignore
  return navigator.deviceMemory || 4
}
