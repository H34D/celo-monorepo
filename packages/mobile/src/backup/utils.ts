import * as _ from 'lodash'
import { AsyncStorage } from 'react-native'
import { generateMnemonic, wordlists } from 'react-native-bip39'
import { getKey, setKey } from 'src/utils/keyStore'
import Logger from 'src/utils/Logger'

const TAG = 'Backup/utils'
const MNEMONIC_SPLITS = 2

export async function createQuizWordList(mnemonic: string, language: string | null) {
  const disallowedWordSet = new Set(mnemonic.split(' '))
  const languageWordList = getWordlist(language)
  const wordOptions: string = await generateMnemonic(1000, null, languageWordList)
  return wordOptions.split(' ').filter((word: string) => !disallowedWordSet.has(word))
}

export function selectQuizWordOptions(
  mnemonic: string,
  allWords: string[],
  numOptions: number
): [string, string[]] | [] {
  const correctWord = _.sample(mnemonic.split(' '))

  if (!correctWord) {
    // mnemonic is empty
    return []
  }

  const wordOptions = _.chain(allWords)
    .sampleSize(numOptions - 1)
    .push(correctWord)
    .shuffle()
    .value()

  return [correctWord, wordOptions]
}

export function getWordlist(language: string | null): string[] {
  switch (language) {
    case 'es': {
      return wordlists.ES
    }
    default: {
      return wordlists.EN
    }
  }
}

function getPrefixWords(wordlist: string[], numWords: number): string[] {
  // Use random words in sorted order for split phrase prefixes. While BIP39
  // does not avoid repeating words, the prefixes MUST be unique otherwise it is
  // not possible to differentiate parts.  Prefixes are just used as a way to
  // determine which mnemonic shard corresponds to which half
  const prefixes = _.chain(wordlist)
    .sampleSize(numWords)
    .uniq()
    .value()
    .sort()

  if (prefixes.length < numWords) {
    throw new Error('Word list has duplicate words')
  }

  return prefixes
}

export function splitMnemonic(mnemonic: string, language: string | null): string[] {
  const mnemonicWords = mnemonic.split(' ')

  const wordlist = getWordlist(language)
  const prefixes = getPrefixWords(wordlist, MNEMONIC_SPLITS)

  const chunkSize = Math.ceil(mnemonicWords.length / MNEMONIC_SPLITS)
  return _.chunk(mnemonicWords, chunkSize).map((words, i) => [prefixes[i], ...words].join(' '))
}

// Sort function based on the first word in string arrays. Mnemonic prefixes are
// the first word in each shard
function sortStringArray(a: string[], b: string[]): number {
  // localeCompare is slower -- https://jsperf.com/operator-vs-localecompage/3
  if (a[0] < b[0]) {
    return -1
  }

  if (a[0] > b[0]) {
    return 1
  }

  return 0
}

export function joinMnemonic(mnemonicShards: string[]) {
  return mnemonicShards
    .map((shard) => shard.split(' '))
    .sort(sortStringArray)
    .map((shard) => shard.slice(1).join(' '))
    .join(' ')
}

// TODO(Rossy) Remove after the next alfa testnet reset
export async function getStoredMnemonic(): Promise<string | null> {
  try {
    Logger.debug(TAG, 'Checking keystore for mnemonic')
    let mnemonic = await getKey('mnemonic')
    if (mnemonic) {
      return mnemonic
    }

    Logger.debug(TAG, 'Mnemonic not found in keystore, checking async storage')
    mnemonic = await AsyncStorage.getItem('mnemonic')
    if (mnemonic) {
      await setKey('mnemonic', mnemonic)
      await AsyncStorage.removeItem('mnemonic')
      return mnemonic
    }

    Logger.error(TAG, 'No mnemonic found')
    return null
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}
