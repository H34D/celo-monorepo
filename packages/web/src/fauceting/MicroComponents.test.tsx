import { ContextualInfo } from 'src/fauceting/MicroComponents'
import * as renderer from 'react-test-renderer'
import { RequestState } from './utils'
import * as React from 'react'

function translator(key: string) {
  return key
}

describe('ContextualInfo', () => {
  it('renders', () => {
    const tree = renderer.create(
      <ContextualInfo requestState={RequestState.Initial} isFaucet={true} t={translator} />
    )
    expect(tree).toMatchSnapshot()
  })
})
