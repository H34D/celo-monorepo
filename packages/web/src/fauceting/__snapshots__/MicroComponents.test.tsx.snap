// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextualInfo renders 1`] = `
<div
  className="css-text-76zvg2"
  dir="auto"
  style={
    Object {
      "color": "rgba(46,51,56,1.00)",
      "fontFamily": "futura-pt, sans-serif",
      "fontSize": "14px",
      "textRendering": "geometricPrecision",
    }
  }
>
  eg. a0000aaa00a0000...a00a0a0000a00a00aa
</div>
`;
